src/auth/src/credentials/mds_credential.rs

```diff
@@ -12,38 +12,154 @@
 }
 
 #[derive(Debug)]
 struct MDSCredential<T>
 where
     T: TokenProvider,
 {
+    quota_project_id: Option<String>,
+    universe_domain: Option<String>,
     token_provider: T,
 }
 
+/// Creates [Credential] instances backed by the [Metadata Service].
+///
+/// While the Google Cloud client libraries for Rust default to credentials
+/// backed by the metadata service, some applications may need to:
+/// * Customize the metadata service credentials in some way
+/// * Bypass the [Application Default Credentials] lookup and only
+///    use the metadata server credentials
+/// * Use the credentials directly outside the client libraries
+///
+/// [Application Default Credentials]: https://cloud.google.com/docs/authentication/application-default-credentials
+/// [Metadata Service]: https://cloud.google.com/compute/docs/metadata/overview
+#[derive(Debug, Default)]
+pub struct MDSCredentialBuilder {
+    endpoint: Option<String>,
+    quota_project_id: Option<String>,
+    scopes: Option<Vec<String>>,
+    universe_domain: Option<String>,
+}
+
+impl MDSCredentialBuilder {
+    /// Sets the endpoint for this credential.
+    /// If not set, the credentials use `http://metadata.google.internal/`.
+    pub fn endpoint<S: Into<String>>(mut self, endpoint: S) -> Self {
+        self.endpoint = Some(endpoint.into());
+        self
+    }
+
+    /// Set the [quota project] for this credential.
+    /// In some services, you can use a service account in
+    /// one project for authentication and authorization, and charge
+    /// the usage to a different project. This may require that the
+    /// service account has serviceusage.services.use on the quota project.
+    ///
+    /// [quota project]: https://cloud.google.com/docs/quotas/quota-project
+    pub fn quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    /// Sets the universe_domain for this credential.
+    /// Client libraries use `universe_domain` to determine
+    /// the API endpoints to use for making requests.
+    /// If not set, then credentials use `googleapis.com`.
+    pub fn universe_domain<S: Into<String>>(mut self, universe_domain: S) -> Self {
+        self.universe_domain = Some(universe_domain.into());
+        self
+    }
+
+    /// Sets the [scopes] for this credential.
+    /// Metadata server issues tokens based on the requested scopes.
+    /// If no scopes are specified, the credential defaults to all
+    /// scopes configured for the [default service account] on the instance.
+    ///
+    /// [default service account]: https://cloud.google.com/iam/docs/service-account-types#default
+    /// [scopes]: https://developers.google.com/identity/protocols/oauth2/scopes
+    pub fn scopes<S: Into<String>>(mut self, scopes: Vec<S>) -> Self {
+        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
+        self
+    }
```

Comment:
Generalize, you want something that supports `IntoIterator<T>`:

```suggestion
    pub fn scopes<I, S>(mut self, scopes: I) -> Self
    where:
        I: IntoIterator<Item = S>,
        S: Into<String>
     {
        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
        self
    }
```